### PR TITLE
RDRP-872- Civil and Defence output

### PR DESCRIPTION
**Pull request by woodac**
## Pull Request submission

Restructured civil & defence output so that it it correlates with the fields in the mapper.

Remove the need for mapper "civil_defence_detailled.csv" so that output is generated from full responses df.

UPDATE (Anne) 
I have added 2 functions that use mocking to test the outputs, I used copilot to do these but am happy with the code generated- I also engineered my prompts using other code in our repo and code already written in this test script.

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes # 872

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

**Review by AnneONS — COMMENTED** [review]
It's looking good so far, though I haven't tested running it yet. Can you think of a way to bring in a unit test, or modify the unit test if one exists? Thanks.

**Review by AnneONS — COMMENTED** [review]
this is really great- we just need to merge in a new version of develop in a little while, then it will be ready to go.

File: src/outputs/intram_by_civil_defence.py [hunk 1]
```diff
@@ -12,46 +12,54 @@ def output_intram_by_civil_defence(
...
+    """
 
+    df_for_output = generate_intram_by_civil_defence(
+        df)
 
-    """
+    # Outputting the CSV file
     output_path = config["outputs_paths"]["outputs_master"]
+    survey_year = config["survey"]["survey_year"]
+    filename = (f"{survey_year}_output_intram_by_civil_defence")
```
Comment (Ryan2Y79) [inline src/outputs/intram_by_civil_defence.py]: "survey_year" is added to the filename on line 35 with the filename being passed to the function `filename_amender` which adds the "survey_year". Therefore line 34 need not add "survey_year" to the filename. Line 34 can be removed totally, and for line 35, name the parameter `filename` equal to `output_intram_by_civil_defence`. You do not need to make a new unit test for your new csv saving function.

File: src/outputs/intram_by_civil_defence.py [hunk 2]
```diff
@@ -12,46 +11,75 @@ def output_intram_by_civil_defence(
...
+        df_for_output (pd.DataFrame): Total intramural expenditure by Civil or Defence
+    """
 
+    df_for_output = generate_intram_by_civil_defence(
+        df)
+
+    # Outputting the CSV file
+    _save_output_intram_civil_def_as_csv(df_for_output, config, write_csv)
+    
+    return df_for_output
```
Comment (AnneONS) [inline src/outputs/intram_by_civil_defence.py]: this doesn't need to be returned, and when the function is called in output_main.py we don't expect something to be retuned.

I suggest return None (and update docstring)

**Review by AnneONS — COMMENTED** [review]
One more thing!

File: src/outputs/intram_by_civil_defence.py (hunk 2 above)
Comment (Ryan2Y79) [reply src/outputs/intram_by_civil_defence.py]: I have addressed this comment.

**Review by adriano-lopresti — COMMENTED** [review]
Happy with the changes to unit tests - I follow them, and they succesfully run

**Review by adriano-lopresti — COMMENTED** [review]
Happy with the changes to unit tests - I follow them, and they succesfully run

**Review by adriano-lopresti — APPROVED** [review]
Happy with the changes to unit tests - I follow them, and they succesfully run

**Pull request merged**